src/instance.rs

```diff
@@ -182,11 +201,76 @@ impl Instance for Wasi {
+        match spec.hooks().as_ref() {
+            None => {
+                log::debug!("no hooks found")
+            }
+            Some(hooks) => {
+                let prestart_hooks = hooks.prestart().as_ref().unwrap();
+
+                for hook in prestart_hooks {
+                    let mut hook_command = process::Command::new(&hook.path());
+                    // Based on OCI spec, the first argument of the args vector is the
+                    // arg0, which can be different from the path.  For example, path
+                    // may be "/usr/bin/true" and arg0 is set to "true". However, rust
+                    // command differenciates arg0 from args, where rust command arg
+                    // doesn't include arg0. So we have to make the split arg0 from the
+                    // rest of args.
+                    if let Some((arg0, args)) = hook.args().as_ref().and_then(|a| a.split_first()) {
+                        log::debug!("run_hooks arg0: {:?}, args: {:?}", arg0, args);
+                        hook_command.arg0(arg0).args(args)
+                    } else {
+                        hook_command.arg0(&hook.path().display().to_string())
+                    };
+
+                    let envs: HashMap<String, String> = if let Some(env) = hook.env() {
+                        parse_env(env)
+                    } else {
+                        HashMap::new()
+                    };
+                    log::debug!("run_hooks envs: {:?}", envs);
+
+                    let mut hook_process = hook_command
+                        .env_clear()
+                        .envs(envs)
+                        .stdin(process::Stdio::piped())
+                        .spawn()
+                        .with_context(|| "Failed to execute hook")?;
+                    let hook_process_pid = Pid::from_raw(hook_process.id() as i32);
+
+                    if let Some(stdin) = &mut hook_process.stdin {
+                        // We want to ignore BrokenPipe here. A BrokenPipe indicates
+                        // either the hook is crashed/errored or it ran successfully.
+                        // Either way, this is an indication that the hook command
+                        // finished execution.  If the hook command was successful,
+                        // which we will check later in this function, we should not
+                        // fail this step here. We still want to check for all the other
+                        // error, in the case that the hook command is waiting for us to
+                        // write to stdin.
+                        let state = format!("{{ \"pid\": {} }}", std::process::id());
+                        let encoded_state =
+                            serde_json::to_string(&state).context("failed to encode container state")?;
+                        if let Err(e) = stdin.write_all(encoded_state.as_bytes()) {
```

Comment:
We actually don't need the encoded state here, the `state` is already a JSON string.
```suggestion
                        if let Err(e) = stdin.write_all(state.as_bytes()) {
```

Reply:
Fixed. Thank you!